Histogram range editor

annehaley opened the pull request:
Resolves #1117, resolves #1109, resolves #1122

Histograms include a mode to "exclude distribution tails". When this mode is enabled, the inputs for min and max become percentages from the ends. These automatically chop off 1% from each end (but if a bucket includes more than 1%, it may end up as more - in the below example, the minimum excludes 2%). The user can still change these values.
![image](https://github.com/girder/large_image/assets/44912689/e3fc2779-2ce9-4def-9329-e97eafa7e652)


Histograms are drawn with some logic to shorten the height of the max frequency if it eclipses all other values.

From @manthey via Slack:

> let v0 be the highest value in the histogram and v1 be the second highest.  Make the top of the histogram min(v0, v1 * 1.5), so that the second highest value is 2/3 the total height unless making it so would make the highest value less than the total height.

An example without max frequency shortening:
![image](https://github.com/girder/large_image/assets/44912689/14abaa79-cbb5-4c58-a3bb-cc37e7d8fb4b)

The same histograms drawn with max frequency shortening:
![image](https://github.com/girder/large_image/assets/44912689/88643a6a-fcc9-42e5-87cd-acd1496ae809)

manthey commented on girder/girder_large_image/web_client/vue/components/FrameSelector.vue:
```diff
@@ -172,10 +174,27 @@ export default Vue.extend({
-        <div v-if="currentModeId > 1" class="image-frame-simple-control">
+        <!-- Use composite layers component twice so state for each one is maintained while invisible -->
+        <!-- Use styling instead of v-if to make each invisible so that the components are not unmounted -->
+        <div  class="image-frame-simple-control" style="width: 500px">
```
Please don't use inline styles -- they are hard to override when we reuse the component elsewhere  (like in HistomicsUI)

annehaley replied:
Oh thanks, I hadn't meant to leave that in there. I wanted to see how it would look when squished for HistomicsUI. Removed in [4f10794](https://github.com/girder/large_image/pull/1175/commits/4f10794f9e22a0bf795e7c544b2af064834579fd)

manthey commented on girder/girder_large_image/web_client/vue/components/CompositeLayers.vue:
```diff
@@ -57,6 +58,44 @@ export default {
+                            frame: this.currentFrame + this.compositeLayerInfo[layer].framedelta,
+                            width: 1024,
+                            height: 1024,
+                            bins: 512,
```
We also need `resample: false` in the query, or the scaling step may end up with histograms on 8-bit data even when the source is not 8-bit.  We might want to have a single place for these parameters rather than repeat them in the bands section.

annehaley replied:
Changed in [b44d581](https://github.com/girder/large_image/pull/1175/commits/b44d581d3a457415036a8f2d46bbcb8d8e2dfbe0)

manthey commented:
I see a lot of extraneous parameters in the tile style queries (layerName, custom, defaultMin, etc).

Also, it doesn't seem to switch to the composited image until I fiddle with a setting -- I was expecting the previous behavior where all of the channels are composited without any specified min/max values until some value has been specified.  This means that the range we display should probably be based purely on the dtype until we fiddle with values.

I don't seem to be able to type fractional percentages in the scale range (I type 0.2 and immediately it reverts to 0).

manthey commented:
Eventually (in another PR), it'd be nice to be able to specify a default for any channel that has blank min/max values.  That is, by default all channels would start with blank min/max values.  If the overall defaults are set to 0.2% / 0.2% or 100 / 20000, then any channel with blank values would use those values, but channels with specified values would use those specific values.  Clearing the numerical values in a channel would go back to using the default, and clearing the default would mean min / max are unspecified in the style.

manthey commented:
We need to change the color of the grab bars on the histograms -- red and black have no contrast in my vision, which makes the grab bars look like bars on the histogram to me.

The icons for showing the histograms are clunky looking and arranged oddly in my view:
![image](https://github.com/girder/large_image/assets/8781639/950bb8f1-da6c-4281-b153-27ac578324b1)

manthey commented:
Can we make the default auto-range be 0.2% (not 0.02%) and prevent the user from dragging it past 50%?

manthey commented on girder/girder_large_image/web_client/vue/components/CompositeLayers.vue:
```diff
@@ -216,15 +321,116 @@ export default {
+</style>
+
+<style>
+input {
```
These need to be scoped or made more exact with some class; otherwise this affects input fields everywhere.  We probably want to add some class specific to this component.

manthey replied:
Maye just make the rule for `.composite-layer-table input`?

annehaley replied:
Changed in [f2f544a](https://github.com/girder/large_image/pull/1175/commits/f2f544a9fff4295393dffb0cc1dd74da01da0808)

annehaley commented:
I removed the `vue-switches` dependency in [d5b27c4](https://github.com/girder/large_image/pull/1175/commits/d5b27c419ffaedc68ebeb0b575e32df175607e60) and now the tests are passing :+1:

manthey commented on girder/girder_large_image/web_client/vue/components/CompositeLayers.vue:
```diff
@@ -216,15 +321,116 @@ export default {
-    overflow-x: auto;
-    overflow-y: hidden;
+    overflow: auto;
+    max-height: 300px;
```
This makes it surprising to edit the colors, as the color dialog gets hidden near the bottom.  Also, if you expand a row such that it causes the scroll bar to appear, then something about the histogram gets cropped and I can't grab the right control handle:
![image](https://github.com/girder/large_image/assets/8781639/d28cdb77-54b4-407a-b519-17998af1045e)
The red band was visible before expanding all, and here the right handle isn't grabbable  any more.

annehaley replied:
For the color pickers, I made them relative to the table container in [0799c52](https://github.com/girder/large_image/pull/1175/commits/0799c52cf27bf2a1595217fd5be5107b5b906720). For the scrollbar/canvas resize conflict, my solution was to make the scrollbar always visible: [35601fc](https://github.com/girder/large_image/pull/1175/commits/35601fc8de52546db289484ca68a8588c229ff63)

manthey commented on girder/girder_large_image/web_client/vue/components/FrameSelector.vue:
```diff
@@ -83,7 +84,7 @@ export default Vue.extend({
                     (b, i) => {
                         if (!b.interpretation) {
-                            return `Band ${i}`
+                            return `Band ${i + 1}`
```
This change could be in a separate PR.

annehaley replied:
Should I take out the `+1`s here?

manthey replied:
No.  It doesn't have to be in a different PR.

manthey replied:
The main difference with https://github.com/girder/large_image/pull/1214 is we now know for all sources how many bands there are and can name them when there are <= 4 bands.

manthey commented:
When I clear the entry in the min or max edit fields, I see a console error: `Error: <text> attribute x: Expected length, "NaN".`

manthey commented:
When I try to show all the histograms of a many channel image, it is asking for the wrong frame for the histogram.  Rather than asking for frames 0,1,2,3,4,... it is asking for 0,1,3,6,10,...  My guess is that frame and framedelta are both set.

annehaley commented:
> When I clear the entry in the min or max edit fields, I see a console error: Error: <text> attribute x: Expected length, "NaN".

Fixed in [4a46617](https://github.com/girder/large_image/pull/1175/commits/4a466178197dcef9454ebf404db70e21bd0e8412)

> When I try to show all the histograms of a many channel image, it is asking for the wrong frame for the histogram. Rather than asking for frames 0,1,2,3,4,... it is asking for 0,1,3,6,10,... My guess is that frame and framedelta are both set.

Fixed in [448ef26](https://github.com/girder/large_image/pull/1175/commits/448ef26b36a59a505ec8006faea88f74e1dfabbe)

manthey commented:
When you have scrolled down the list of channels, the color picker is out of view (scrolling up shows it).

annehaley commented:
> When you have scrolled down the list of channels, the color picker is out of view (scrolling up shows it).

Changed in [3e2fd3a](https://github.com/girder/large_image/pull/1175/commits/3e2fd3a80c275d427dbeff123c96f6a0d2c8e740). Rather than having a color picker component for every row in the table with each being relative to the color picker column, we now have one color picker component relative to the table header, which is sticky during scrolling.

manthey commented on girder/girder_large_image/web_client/vue/components/FrameSelector.vue:
```diff
@@ -33,7 +34,7 @@ export default Vue.extend({
     methods: {
         updateStyle(style) {
-            this.style = style
+            this.style = Object.assign(this.style, style)
```
What does doing `this.style = ` in front of the `Object.assign` do?

manthey reviewed (approved):
Great.
